bootloader/piboot.go

```diff
@@ -0,0 +1,429 @@
+	} else {
+		// install/recovery modes, use recovery kernel
+		prefix = filepath.Join("/systems", env.Get("snapd_recovery_system"),
+			"kernel")
+		cfgDir = p.rootdir
+		dstDir = filepath.Join(p.rootdir, prefix)
+	}
+
+	logger.Debugf("configure piboot %s with prefix %q, cfgDir %q, dstDir %q",
+		cfgFile, prefix, cfgDir, dstDir)
+
+	if err := os.MkdirAll(dstDir, 0755); err != nil {
+		return err
+	}
+	return p.applyConfig(env, cfgFile, prefix, cfgDir, dstDir)
+}
+
+// Sets os_prefix in RPi config.txt or tryboot.txt
+func (p *piboot) setRPiCfgOsPrefix(prefix, inFile, outFile string) error {
+	buf, err := ioutil.ReadFile(inFile)
+	if err != nil {
+		return err
+	}
+
+	lines := strings.Split(string(buf), "\n")
+
+	replaced := false
+	newOsPrefix := "os_prefix=" + prefix + "/"
+	for i, line := range lines {
+		if strings.HasPrefix(line, "os_prefix=") {
+			if replaced {
+				logger.Noticef("unexpected extra os_prefix line: %q", line)
+				lines[i] = "# " + lines[i]
+				continue
+			}
+			lines[i] = newOsPrefix
+			replaced = true
+		}
+	}
+	if !replaced {
+		lines = append(lines, newOsPrefix)
+		lines = append(lines, "")
+	}
+
+	output := strings.Join(lines, "\n")
+	return osutil.AtomicWriteFile(outFile, []byte(output), 0644, 0)
+}
+
+func (p *piboot) createCmdline(env *ubootenv.Env, defaultsFile, outFile string) error {
+	var base string
+	fullCmdline := env.Get("snapd_full_cmdline_args")
+	if fullCmdline != "" {
+		base = fullCmdline
+	} else {
+		buf, err := ioutil.ReadFile(defaultsFile)
+		if err != nil {
+			return err
+		}
+
+		lines := strings.Split(string(buf), "\n")
+		base = lines[0]
+		base += " " + env.Get("snapd_extra_cmdline_args")
+	}
+
+	mode := env.Get("snapd_recovery_mode")
+	cmdline := base + " snapd_recovery_mode=" + mode
+	if mode != "run" {
+		cmdline += " snapd_recovery_system=" + env.Get("snapd_recovery_system")
+	}
+	// Signal when we are trying a new kernel
+	kernelStatus := env.Get("kernel_status")
+	if kernelStatus == "try" {
+		cmdline += " kernel_status=trying"
+	}
+	cmdline += "\n"
+
+	logger.Debugf("writing kernel command line to %s", outFile)
+
+	return osutil.AtomicWriteFile(outFile, []byte(cmdline), 0644, 0)
+}
+
+// Configure pi bootloader with a given os_prefix. cfgDir contains the
+// config files, and dstDir is where we will place the kernel command
+// line.
+func (p *piboot) applyConfig(env *ubootenv.Env,
+	configFile, prefix, cfgDir, dstDir string) error {
+
+	cmdlineFile := filepath.Join(dstDir, "cmdline.txt")
+	refCmdlineFile := filepath.Join(cfgDir, "cmdline.txt")
+	currentConfigFile := filepath.Join(cfgDir, "config.txt")
+
+	if err := p.createCmdline(env, refCmdlineFile, cmdlineFile); err != nil {
+		return err
+	}
+	if err := p.setRPiCfgOsPrefix(prefix, currentConfigFile,
+		filepath.Join(cfgDir, configFile)); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (p *piboot) GetBootVars(names ...string) (map[string]string, error) {
+	out := make(map[string]string)
```

Comment:
nitpick: I would move this under the `OpenWithFlags` call, and specify the capacity to `len(names)`

Reply:
Good suggestion, I've done this now